### PR TITLE
fix(proctypes): update service after all the proctypes are deployed

### LINKED ISSUE #994: Conflict updating service during deploy with multiple process types
The deployment of a rails buildpack app with the default process types (web, rake, worker, console) sporadically fails with a conflict when it tries to update the service after the build and kubernetes deploy has finished.

The app has it's web process type scaled to 2, everything else is at 0 and is using kubernetes deployments.

I am running controller v2.4.0 on workflow v2.4.0 with k8s v1.3.4 in a multi-az setup with one node in each zone.

Controller Log:

```
10.2.52.11 "GET /v2/hooks/key/c4:50:c8:06:f7:76:1b:a3:1f:3d:56:ce:75:72:db:65 HTTP/1.1" 200 137 "deis-builder"
10.2.52.11 "POST /v2/hooks/config/ HTTP/1.1" 200 3113 "deis-builder"
INFO [shop-backend]: build shop-backend-0e9d0d8 created
INFO [shop-backend]: gitlab-runner deployed 0da5e09
INFO [shop-backend]: buildpack type detected. Defaulting to $PORT 5000
INFO [shop-backend]: buildpack type detected. Defaulting to $PORT 5000
INFO [shop-backend]: buildpack type detected. Defaulting to $PORT 5000
INFO [shop-backend]: buildpack type detected. Defaulting to $PORT 5000
INFO [shop-backend]: adding 30s on to the original 120s timeout to account for the initial delay specified in the liveness / readiness probe
INFO [shop-backend]: This deployments overall timeout is 150s - batch timout is 150s and there are 1 batches to deploy with a total of 2 pods
INFO [shop-backend]: waited 10s and 2 pods are in service
INFO [shop-backend]: waited 20s and 2 pods are in service
INFO [shop-backend]: waited 30s and 2 pods are in service
ERROR [shop-backend]: (app::deploy): ('failed to update Service "shop-backend" in Namespace "shop-backend": 409 Conflict', 'shop-backend', 'shop-backend')
ERROR:root:(app::deploy): ('failed to update Service "shop-backend" in Namespace "shop-backend": 409 Conflict', 'shop-backend', 'shop-backend')
Traceback (most recent call last):
  File "/app/scheduler/__init__.py", line 206, in _update_application_service
    self.update_service(namespace, namespace, data=service)
  File "/app/scheduler/__init__.py", line 1219, in update_service
    'update Service "{}" in Namespace "{}"', namespace, name
scheduler.KubeHTTPException: ('failed to update Service "shop-backend" in Namespace "shop-backend": 409 Conflict', 'shop-backend', 'shop-backend')
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/app/api/models/app.py", line 562, in deploy
    async_run(tasks)
  File "/app/api/utils.py", line 169, in async_run
    raise error
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "/app/api/utils.py", line 182, in async_task
    yield from loop.run_in_executor(None, params)
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/app/scheduler/__init__.py", line 138, in deploy
    self._update_application_service(namespace, name, app_type, port, routable, service_annotations)  # noqa
  File "/app/scheduler/__init__.py", line 209, in _update_application_service
    self.update_service(namespace, namespace, data=old_service)
  File "/app/scheduler/__init__.py", line 1219, in update_service
    'update Service "{}" in Namespace "{}"', namespace, name
scheduler.KubeHTTPException: ('failed to update Service "shop-backend" in Namespace "shop-backend": 409 Conflict', 'shop-backend', 'shop-backend')
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/app/api/models/build.py", line 64, in create
    self.app.deploy(new_release)
  File "/app/api/models/app.py", line 566, in deploy
    raise ServiceUnavailable(err) from e
api.exceptions.ServiceUnavailable: (app::deploy): ('failed to update Service "shop-backend" in Namespace "shop-backend": 409 Conflict', 'shop-backend', 'shop-backend')
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/views.py", line 471, in dispatch
    response = handler(request, *args, **kwargs)
  File "/app/api/views.py", line 486, in create
    super(BuildHookViewSet, self).create(request, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/mixins.py", line 21, in create
    self.perform_create(serializer)
  File "/app/api/viewsets.py", line 21, in perform_create
    self.post_save(obj)
  File "/app/api/views.py", line 492, in post_save
    build.create(self.user)
  File "/app/api/models/build.py", line 71, in create
    raise DeisException(str(e)) from e
api.exceptions.DeisException: (app::deploy): ('failed to update Service "shop-backend" in Namespace "shop-backend": 409 Conflict', 'shop-backend', 'shop-backend')
10.2.52.11 "POST /v2/hooks/build/ HTTP/1.1" 400 149 "deis-builder"
```

### LINKED ISSUE #999: `git push` throwing error when controller tries to modify annotations
```
$ g push deis master -f
Counting objects: 1, done.
Writing objects: 100% (1/1), 267 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 1 (delta 0)
Starting build... but first, coffee!
-----> Ruby app detected
-----> Compiling Ruby/Rack
-----> Using Ruby version: ruby-2.0.0
-----> Installing dependencies using bundler 1.11.2
       Running: bundle install --without development:test --path vendor/bundle --binstubs vendor/bundle/bin -j4 --deployment
       Fetching gem metadata from http://rubygems.org/..........
       Fetching version metadata from http://rubygems.org/..
       Rubygems 2.0.14.1 is not threadsafe, so your gems will be installed one at a time. Upgrade to Rubygems 2.1.0 or higher to enable parallel gem installation.
       Installing rack 1.6.1
       Installing tilt 2.0.1
       Using bundler 1.11.2
       Installing rack-protection 1.5.3
       Installing sinatra 1.4.6
       Bundle complete! 1 Gemfile dependency, 5 gems now installed.
       Gems in the groups development and test were not installed.
       Bundled gems are installed into ./vendor/bundle.
       Bundle completed (2.15s)
       Cleaning up the bundler cache.

-----> Discovering process types
       Procfile declares types -> web
       Default process types for Ruby -> rake, console, web
-----> Compiled slug size is 16M
Build complete.
Launching App...
...
...
...
remote: 2016/08/21 05:15:47 Error running git receive hook [publishing release (Unknown Error (400): {"detail":"(app::deploy): 'annotations'"})]
To ssh://git@deis-builder.nitro.deis-demo.works:2222/ribbed-pachinko.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'ssh://git@deis-builder.nitro.deis-demo.works:2222/ribbed-pachinko.git'
```

```
INFO [ribbed-pachinko]: build ribbed-pachinko-a67ef61 created
INFO:api.models.app:[ribbed-pachinko]: build ribbed-pachinko-a67ef61 created
INFO [ribbed-pachinko]: jhansen deployed e159088
INFO:api.models.app:[ribbed-pachinko]: jhansen deployed e159088
INFO [ribbed-pachinko]: buildpack type detected. Defaulting to $PORT 5000
INFO:api.models.app:[ribbed-pachinko]: buildpack type detected. Defaulting to $PORT 5000
INFO [ribbed-pachinko]: adding 30s on to the original 120s timeout to account for the initial delay specified in the liveness / readiness probe
INFO:scheduler:[ribbed-pachinko]: adding 30s on to the original 120s timeout to account for the initial delay specified in the liveness / readiness probe
INFO [ribbed-pachinko]: This deployments overall timeout is 150s - batch timout is 150s and there are 1 batches to deploy with a total of 1 pods
INFO:scheduler:[ribbed-pachinko]: This deployments overall timeout is 150s - batch timout is 150s and there are 1 batches to deploy with a total of 1 pods
INFO [ribbed-pachinko]: waited 10s and 1 pods are in service
INFO:scheduler:[ribbed-pachinko]: waited 10s and 1 pods are in service
INFO [ribbed-pachinko]: waited 20s and 1 pods are in service
INFO:scheduler:[ribbed-pachinko]: waited 20s and 1 pods are in service
INFO [ribbed-pachinko]: waited 30s and 1 pods are in service
INFO:scheduler:[ribbed-pachinko]: waited 30s and 1 pods are in service
ERROR [ribbed-pachinko]: (app::deploy): 'annotations'
ERROR:api.models.app:[ribbed-pachinko]: (app::deploy): 'annotations'
ERROR:root:(app::deploy): 'annotations'
Traceback (most recent call last):
  File "/app/scheduler/__init__.py", line 198, in _update_application_service
    service['metadata']['annotations']['router.deis.io/%s' % key] = str(value)
KeyError: 'annotations'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/api/models/app.py", line 562, in deploy
    async_run(tasks)
  File "/app/api/utils.py", line 169, in async_run
    raise error
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "/app/api/utils.py", line 182, in async_task
    yield from loop.run_in_executor(None, params)
  File "/usr/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/app/scheduler/__init__.py", line 148, in deploy
    self._update_application_service(namespace, name, app_type, port, routable, service_annotations)  # noqa
  File "/app/scheduler/__init__.py", line 220, in _update_application_service
    raise KubeException(str(e)) from e
scheduler.KubeException: 'annotations'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/api/models/build.py", line 64, in create
    self.app.deploy(new_release)
  File "/app/api/models/app.py", line 566, in deploy
    raise ServiceUnavailable(err) from e
api.exceptions.ServiceUnavailable: (app::deploy): 'annotations'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/views.py", line 471, in dispatch
    response = handler(request, *args, **kwargs)
  File "/app/api/views.py", line 486, in create
    super(BuildHookViewSet, self).create(request, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/mixins.py", line 21, in create
    self.perform_create(serializer)
  File "/app/api/viewsets.py", line 21, in perform_create
    self.post_save(obj)
  File "/app/api/views.py", line 492, in post_save
    build.create(self.user)
  File "/app/api/models/build.py", line 71, in create
    raise DeisException(str(e)) from e
api.exceptions.DeisException: (app::deploy): 'annotations'
10.2.87.4 "POST /v2/hooks/build/ HTTP/1.1" 400 41 "deis-builder"
w:carbon k:nitro example-ruby-sinatra [master]$
```

```
w:carbon k:nitro example-ruby-sinatra [master]$ k --namespace=ribbed-pachinko get svc,deployments,rs,po
NAME                                   CLUSTER-IP   EXTERNAL-IP   PORT(S)      AGE
ribbed-pachinko                        10.3.0.168   <none>        80/TCP       8m
NAME                                   DESIRED      CURRENT       UP-TO-DATE   AVAILABLE   AGE
ribbed-pachinko-web                    1            1             1            1           8m
NAME                                   DESIRED      CURRENT       AGE
ribbed-pachinko-web-1330408578         0            0             8m
ribbed-pachinko-web-2883480714         1            1             2m
NAME                                   READY        STATUS        RESTARTS   AGE
ribbed-pachinko-web-2883480714-4obn5   1/1          Running       0          2m
w:carbon k:nitro example-ruby-sinatra [master]$ k --namespace=ribbed-pachinko describe svc ribbed-pachinko
Name:			ribbed-pachinko
Namespace:		ribbed-pachinko
Labels:			app=ribbed-pachinko,heritage=deis
Selector:		app=ribbed-pachinko,heritage=deis
Type:			ClusterIP
IP:			10.3.0.168
Port:			http	80/TCP
Endpoints:		10.2.47.9:5000
Session Affinity:	None
No events.
```

### PR DESCRIPTION
**Pull request by kmala**
fixes #994 
fixes #999

### CODE REVIEW

File: rootfs/api/models/app.py [hunk 1]
```diff
@@ -565,15 +554,36 @@ def deploy(self, release, force_deploy=False):
...
-        # Wait until application is available in the router
-        # Only run when there is no previous build / release
-        old = release.previous()
-        if old is None or old.build is None:
-            self.verify_application_health(**kwargs)
+        app_type = 'web' if 'web' in deploys else 'cmd' if 'cmd' in deploys else None
+        # Make sure the application is routable and uses the correct port Done after the fact to
+        # let initial deploy settle before routing traffic to the application
+        if deploys and app_type:
+            self._update_application_service(self.id, app_type, deploys[app_type].get('envs', {}).get('PORT', None), deploys[app_type].get('routable'), service_annotations)  # noqa
```
Comment (helgi) [inline rootfs/api/models/app.py]: Can this be split up? Would make it more readable

File: rootfs/api/models/app.py [hunk 2]
```diff
@@ -565,15 +554,38 @@ def deploy(self, release, force_deploy=False):
...
+            old = release.previous()
+            if old is None or old.build is None:
+                self.verify_application_health(**deploys[app_type])
 
         # cleanup old release objects from kubernetes
         release.cleanup_old()
 
+    def _check_deployment_in_progress(self, deploys, force_deploy=False):
+        for scale_type, kwargs in deploys.items():
+            if force_deploy:
```
Comment (mboersma) [inline rootfs/api/models/app.py]: If `force_deploy` isn't changed in this function, then it's a loop invariant and can be moved outside the loop. (Python's compiler isn't smart enough to do that for you.)
```python
if force_deploy:
    return

for scale_type, kwargs in deploys.items():
    # Is there an existing deployment in progress?
    name = self._get_job_id(scale_type)
...
```

File: rootfs/api/models/app.py [hunk 3]
```diff
@@ -869,3 +881,35 @@ def routable(self, routable):
         except KubeException as e:
             self._scheduler.update_service(self.id, self.id, data=old_service)
             raise ServiceUnavailable(str(e)) from e
+
+    def _update_application_service(self, namespace, app_type, port, routable=False, annotations={}):  # noqa
```
Comment (mboersma) [inline rootfs/api/models/app.py]: Why `# noqa` here? Did it fail the McCabe complexity check?

Comment (kmala) [inline rootfs/api/models/app.py]: the flake check failed for length.

Comment (helgi) [inline rootfs/api/models/app.py]: I'm going to let @kmala answer that one but having `{}` as a default is generally a failure for the linter. Supposedly you are supposed to use `None` and then do other stuff internally... kinda funky

File: rootfs/api/models/app.py (hunk 2 above)
Comment (kmala) [inline rootfs/api/models/app.py]: yeah..i just blindly copied during the refactor.Changed it.

**Pull request merged**